File: rental/src/test/java/io/axoniq/demo/bikerental/rental/command/BikeTest.java
package io.axoniq.demo.bikerental.rental.command;

import io.axoniq.demo.bikerental.coreapi.rental.ApproveRequestCommand;
import io.axoniq.demo.bikerental.coreapi.rental.BikeInUseEvent;
import io.axoniq.demo.bikerental.coreapi.rental.BikeRegisteredEvent;
import io.axoniq.demo.bikerental.coreapi.rental.BikeRequestedEvent;
import io.axoniq.demo.bikerental.coreapi.rental.BikeReturnedEvent;
import io.axoniq.demo.bikerental.coreapi.rental.RegisterBikeCommand;
import io.axoniq.demo.bikerental.coreapi.rental.RejectRequestCommand;
import io.axoniq.demo.bikerental.coreapi.rental.RequestBikeCommand;
import io.axoniq.demo.bikerental.coreapi.rental.RequestRejectedEvent;
import io.axoniq.demo.bikerental.coreapi.rental.ReturnBikeCommand;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Plain unit tests for the Bike decision model. We build the model by replaying historical events
 * via {@link Bike#apply(Object)} (the equivalent of "given") and then exercise a decision method
 * (the equivalent of "when"), asserting on the events produced.
 */
class BikeTest {

    private static final String REFERENCE = "rentalId";

    private Bike given(Object... events) {
        Bike bike = new Bike();
        for (Object event : events) {
            bike.apply(event);
        }
        return bike;
    }

    @Test
    void canRegisterBike() {
        // decideOnRegister carries a deliberate time-based "taint"; retry across the 5s window.
        List<Object> events = registerWithRetry(new RegisterBikeCommand("bikeId-1234", "city-bike", "Amsterdam"));
        assertEquals(List.of(new BikeRegisteredEvent("bikeId-1234", "city-bike", "Amsterdam")), events);
    }

    @Test
    void cannotRegisterExistingBike() {
        Bike bike = given(new BikeRegisteredEvent("bikeId", "city", "Amsterdam"));
        assertThrows(IllegalStateException.class,
                     () -> bike.decideOnRegister(new RegisterBikeCommand("bikeId", "city", "Amsterdam")));
    }

    @Test
    void canRequestAvailableBike() {
        Bike bike = given(new BikeRegisteredEvent("bikeId", "city", "Amsterdam"));
        List<Object> events = bike.decideOnRequest(new RequestBikeCommand("bikeId", "rider"), REFERENCE);
        assertEquals(1, events.size());
        BikeRequestedEvent event = assertInstanceOf(BikeRequestedEvent.class, events.get(0));
        assertEquals("bikeId", event.bikeId());
        assertEquals("rider", event.renter());
    }

    @Test
    void cannotRequestAlreadyRequestedBike() {
        Bike bike = given(new BikeRegisteredEvent("bikeId", "city", "Amsterdam"),
                          new BikeRequestedEvent("bikeId", "rider", REFERENCE));
        assertThrows(IllegalStateException.class,
                     () -> bike.decideOnRequest(new RequestBikeCommand("bikeId", "rider"), REFERENCE));
    }

    @Test
    void canApproveRequestedBike() {
        Bike bike = given(new BikeRegisteredEvent("bikeId", "city", "Amsterdam"),
                          new BikeRequestedEvent("bikeId", "rider", REFERENCE));
        assertEquals(List.of(new BikeInUseEvent("bikeId", "rider")),
                     bike.decideOnApprove(new ApproveRequestCommand("bikeId", "rider")));
    }

    @Test
    void canRejectRequestedBike() {
        Bike bike = given(new BikeRegisteredEvent("bikeId", "city", "Amsterdam"),
                          new BikeRequestedEvent("bikeId", "rider", REFERENCE));
        assertEquals(List.of(new RequestRejectedEvent("bikeId")),
                     bike.decideOnReject(new RejectRequestCommand("bikeId", "rider")));
    }

    @Test
    void cannotRejectRequestForWrongRequester() {
        Bike bike = given(new BikeRegisteredEvent("bikeId", "city", "Amsterdam"),
                          new BikeRequestedEvent("bikeId", "rider", REFERENCE));
        assertTrue(bike.decideOnReject(new RejectRequestCommand("bikeId", "otherRider")).isEmpty());
    }

    @Test
    void cannotApproveRequestForAnotherRider() {
        Bike bike = given(new BikeRegisteredEvent("bikeId", "city", "Amsterdam"),
                          new BikeRequestedEvent("bikeId", "rider", REFERENCE));
        assertTrue(bike.decideOnApprove(new ApproveRequestCommand("bikeId", "otherRider")).isEmpty());
    }

    @Test
    void canReturnBikeInUse() {
        Bike bike = given(new BikeRegisteredEvent("bikeId", "city", "Amsterdam"),
                          new BikeRequestedEvent("bikeId", "rider", REFERENCE),
                          new BikeInUseEvent("bikeId", "rider"));
        assertEquals(List.of(new BikeReturnedEvent("bikeId", "NewLocation")),
                     bike.decideOnReturn(new ReturnBikeCommand("bikeId", "NewLocation")));
    }

    @Test
    void cannotRequestBikeInUse() {
        Bike bike = given(new BikeRegisteredEvent("bikeId", "city", "Amsterdam"),
                          new BikeRequestedEvent("bikeId", "rider", REFERENCE),
                          new BikeInUseEvent("bikeId", "rider"));
        assertThrows(IllegalStateException.class,
                     () -> bike.decideOnRequest(new RequestBikeCommand("bikeId", "otherRenter"), REFERENCE));
    }

    @Test
    void canRequestReturnedBike() {
        Bike bike = given(new BikeRegisteredEvent("bikeId", "city", "Amsterdam"),
                          new BikeRequestedEvent("bikeId", "rider", REFERENCE),
                          new BikeInUseEvent("bikeId", "rider"),
                          new BikeReturnedEvent("bikeId", "NewLocation"));
        List<Object> events = bike.decideOnRequest(new RequestBikeCommand("bikeId", "newRider"), REFERENCE);
        BikeRequestedEvent event = assertInstanceOf(BikeRequestedEvent.class, events.get(0));
        assertEquals("newRider", event.renter());
    }

    @Test
    void canRequestRejectedBike() {
        Bike bike = given(new BikeRegisteredEvent("bikeId", "city", "Amsterdam"),
                          new BikeRequestedEvent("bikeId", "rider", REFERENCE),
                          new RequestRejectedEvent("bikeId"));
        List<Object> events = bike.decideOnRequest(new RequestBikeCommand("bikeId", "newRider"), REFERENCE);
        BikeRequestedEvent event = assertInstanceOf(BikeRequestedEvent.class, events.get(0));
        assertEquals("newRider", event.renter());
    }

    private List<Object> registerWithRetry(RegisterBikeCommand command) {
        for (int attempt = 0; attempt < 6; attempt++) {
            try {
                return new Bike().decideOnRegister(command);
            } catch (IllegalStateException e) {
                try {
                    Thread.sleep(1100);
                } catch (InterruptedException interrupted) {
                    Thread.currentThread().interrupt();
                    throw new IllegalStateException(interrupted);
                }
            }
        }
        throw new IllegalStateException("Could not register a bike within the retry window");
    }
}

File: rental/src/command/bike.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';
import {
  ApproveRequestCommand,
  BikeInUseEvent,
  BikeRegisteredEvent,
  BikeRequestedEvent,
  BikeReturnedEvent,
  RegisterBikeCommand,
  RejectRequestCommand,
  RequestBikeCommand,
  RequestRejectedEvent,
  ReturnBikeCommand,
} from '@bikerental/core-api';
import { Bike } from './bike.js';

const REFERENCE = 'rentalId';

const delay = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

function given(...events: object[]): Bike {
  const bike = new Bike();
  for (const event of events) {
    bike.apply(event);
  }
  return bike;
}

// decideOnRegister carries a deliberate time-based "taint"; retry across the 5s window.
async function registerWithRetry(command: RegisterBikeCommand): Promise<object[]> {
  for (let attempt = 0; attempt < 6; attempt++) {
    try {
      return new Bike().decideOnRegister(command);
    } catch {
      await delay(1100);
    }
  }
  throw new Error('Could not register a bike within the retry window');
}

test('canRegisterBike', async () => {
  const events = await registerWithRetry(new RegisterBikeCommand('bikeId-1234', 'city-bike', 'Amsterdam'));
  assert.deepStrictEqual(events, [new BikeRegisteredEvent('bikeId-1234', 'city-bike', 'Amsterdam')]);
});

test('cannotRegisterExistingBike', () => {
  const bike = given(new BikeRegisteredEvent('bikeId', 'city', 'Amsterdam'));
  assert.throws(() => bike.decideOnRegister(new RegisterBikeCommand('bikeId', 'city', 'Amsterdam')));
});

test('canRequestAvailableBike', () => {
  const bike = given(new BikeRegisteredEvent('bikeId', 'city', 'Amsterdam'));
  const events = bike.decideOnRequest(new RequestBikeCommand('bikeId', 'rider'), REFERENCE);
  assert.equal(events.length, 1);
  const event = events[0] as BikeRequestedEvent;
  assert.ok(event instanceof BikeRequestedEvent);
  assert.equal(event.bikeId, 'bikeId');
  assert.equal(event.renter, 'rider');
});

test('cannotRequestAlreadyRequestedBike', () => {
  const bike = given(
    new BikeRegisteredEvent('bikeId', 'city', 'Amsterdam'),
    new BikeRequestedEvent('bikeId', 'rider', REFERENCE),
  );
  assert.throws(() => bike.decideOnRequest(new RequestBikeCommand('bikeId', 'rider'), REFERENCE));
});

test('canApproveRequestedBike', () => {
  const bike = given(
    new BikeRegisteredEvent('bikeId', 'city', 'Amsterdam'),
    new BikeRequestedEvent('bikeId', 'rider', REFERENCE),
  );
  assert.deepStrictEqual(bike.decideOnApprove(new ApproveRequestCommand('bikeId', 'rider')), [
    new BikeInUseEvent('bikeId', 'rider'),
  ]);
});

test('canRejectRequestedBike', () => {
  const bike = given(
    new BikeRegisteredEvent('bikeId', 'city', 'Amsterdam'),
    new BikeRequestedEvent('bikeId', 'rider', REFERENCE),
  );
  assert.deepStrictEqual(bike.decideOnReject(new RejectRequestCommand('bikeId', 'rider')), [
    new RequestRejectedEvent('bikeId'),
  ]);
});

test('cannotRejectRequestForWrongRequester', () => {
  const bike = given(
    new BikeRegisteredEvent('bikeId', 'city', 'Amsterdam'),
    new BikeRequestedEvent('bikeId', 'rider', REFERENCE),
  );
  assert.equal(bike.decideOnReject(new RejectRequestCommand('bikeId', 'otherRider')).length, 0);
});

test('cannotApproveRequestForAnotherRider', () => {
  const bike = given(
    new BikeRegisteredEvent('bikeId', 'city', 'Amsterdam'),
    new BikeRequestedEvent('bikeId', 'rider', REFERENCE),
  );
  assert.equal(bike.decideOnApprove(new ApproveRequestCommand('bikeId', 'otherRider')).length, 0);
});

test('canReturnBikeInUse', () => {
  const bike = given(
    new BikeRegisteredEvent('bikeId', 'city', 'Amsterdam'),
    new BikeRequestedEvent('bikeId', 'rider', REFERENCE),
    new BikeInUseEvent('bikeId', 'rider'),
  );
  assert.deepStrictEqual(bike.decideOnReturn(new ReturnBikeCommand('bikeId', 'NewLocation')), [
    new BikeReturnedEvent('bikeId', 'NewLocation'),
  ]);
});

test('cannotRequestBikeInUse', () => {
  const bike = given(
    new BikeRegisteredEvent('bikeId', 'city', 'Amsterdam'),
    new BikeRequestedEvent('bikeId', 'rider', REFERENCE),
    new BikeInUseEvent('bikeId', 'rider'),
  );
  assert.throws(() => bike.decideOnRequest(new RequestBikeCommand('bikeId', 'otherRenter'), REFERENCE));
});

test('canRequestReturnedBike', () => {
  const bike = given(
    new BikeRegisteredEvent('bikeId', 'city', 'Amsterdam'),
    new BikeRequestedEvent('bikeId', 'rider', REFERENCE),
    new BikeInUseEvent('bikeId', 'rider'),
    new BikeReturnedEvent('bikeId', 'NewLocation'),
  );
  const events = bike.decideOnRequest(new RequestBikeCommand('bikeId', 'newRider'), REFERENCE);
  const event = events[0] as BikeRequestedEvent;
  assert.ok(event instanceof BikeRequestedEvent);
  assert.equal(event.renter, 'newRider');
});

test('canRequestRejectedBike', () => {
  const bike = given(
    new BikeRegisteredEvent('bikeId', 'city', 'Amsterdam'),
    new BikeRequestedEvent('bikeId', 'rider', REFERENCE),
    new RequestRejectedEvent('bikeId'),
  );
  const events = bike.decideOnRequest(new RequestBikeCommand('bikeId', 'newRider'), REFERENCE);
  const event = events[0] as BikeRequestedEvent;
  assert.ok(event instanceof BikeRequestedEvent);
  assert.equal(event.renter, 'newRider');
});
